File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/complex/table/TableLine.java
package com.epam.jdi.uitests.mobile.appium.elements.complex.table;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.mobile.appium.elements.base.Element;
import com.epam.jdi.uitests.mobile.appium.elements.base.SelectElement;
import com.epam.jdi.uitests.mobile.appium.elements.complex.table.interfaces.ITableLine;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.epam.commons.LinqUtils.*;
import static com.epam.commons.ReflectionUtils.isClass;
import static com.epam.jdi.uitests.core.settings.JDISettings.asserter;
import static com.epam.jdi.uitests.mobile.appium.driver.WebDriverByUtils.fillByTemplate;

/**
 * Created by 12345 on 25.10.2014.
 */
abstract class TableLine extends Element implements ITableLine, Cloneable {
    public boolean hasHeader;
    public ElementIndexType elementIndex;
    public Table table;
    protected int count = 0;
    protected String[] headers;
    protected int startIndex = 1;
    protected By headersLocator;
    protected By defaultTemplate;
    protected By lineTemplate = null;

    protected <T extends TableLine> T clone(T newTableLine, Table newTable) {
        asserter.silent(() -> super.clone());
        newTableLine.hasHeader = hasHeader;
        newTableLine.elementIndex = elementIndex;
        newTableLine.table = newTable;
        newTableLine.count = count;
        newTableLine.headers = headers;
        newTableLine.startIndex = startIndex;
        newTableLine.headersLocator = headersLocator;
        newTableLine.defaultTemplate = defaultTemplate;
        newTableLine.lineTemplate = lineTemplate;
        return newTableLine;
    }

    public int getStartIndex() {
        return startIndex;
    }

    protected List<WebElement> getLineAction(int colNum) {
        return table.getWebElement().findElements(fillByTemplate((lineTemplate != null) ? lineTemplate : defaultTemplate, colNum));
    }

    protected List<WebElement> getLineAction(String lineName) {
        return (lineTemplate == null)
                ? getLineAction(getIndex(headers(), lineName) + 1)
                : table.getWebElement().findElements(fillByTemplate(lineTemplate, lineName));
    }

    protected abstract List<WebElement> getFirstLine();

    protected abstract int getCount();

    public void setCount(int value) {
        if (table.cache) count = value;
    }

    public int count() {
        return count > 0 ? count
                : headers != null && headers.length > 0 ? headers.length : getCount();
    }

    public void clean() {
        headers = null;
        count = 0;
    }

    public void setHeaders(String[] value) {
        if (table.cache)
            headers = value.clone();
    }

    protected String[] getHeadersTextAction() {
        return toStringArray(select(getHeadersAction(), WebElement::getText));
    }

    protected abstract List<WebElement> getHeadersAction();

    public final MapArray<String, SelectElement> header() {
        return new MapArray<>(getHeadersAction(), WebElement::getText, SelectElement::new);
    }

    public final SelectElement header(String name) {
        return header().get(name);
    }

    public String[] headers() {
        if (headers != null)
            return headers.clone();
        String[] localHeaders = hasHeader
                ? timer().getResult(this::getHeadersTextAction)
                : getNumList(count());
        if (localHeaders == null || localHeaders.length == 0)
            return new String[]{};
        if (count > 0 && localHeaders.length > count)
            localHeaders = toStringArray(Arrays.asList(localHeaders).subList(0, count));
        setHeaders(localHeaders);
        setCount(localHeaders.length);
        return localHeaders;
    }

    protected String[] getNumList(int count) {
        return getNumList(count, 1);
    }

    protected String[] getNumList(int count, int from) {
        List<String> result = new ArrayList<>();
        for (int i = from; i < count + from; i++)
            result.add(Integer.toString(i));
        return result.toArray(new String[count]);
    }

    public final void update(TableLine tableLine) {
        if (tableLine.count > 0)
            setCount(tableLine.count());
        if (tableLine.startIndex != 1)
            startIndex = tableLine.startIndex;
        if (tableLine.headers != null && tableLine.headers.length > 0)
            setHeaders(tableLine.headers());
        if ((isClass(tableLine.getClass(), Columns.class) && !tableLine.hasHeader)
                || (isClass(tableLine.getClass(), Rows.class) && tableLine.hasHeader))
            hasHeader = tableLine.hasHeader;
        if (tableLine.elementIndex != ElementIndexType.Nums)
            elementIndex = tableLine.elementIndex;
    }

    public final MapArray<String, MapArray<String, String>> getAsText() {
        return get().toMapArray(line -> line.toMapArray(IText::getText));
    }
}

File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/Table/TableLine.cs
using System.Collections.Generic;
using System.Linq;
using JDI_Web.Selenium.Base;
using JDI_Web.Selenium.DriverFactory;
using JDI_Web.Selenium.Elements.Base;
using JDI_Web.Selenium.Elements.Complex.Table.Interfaces;
using OpenQA.Selenium;

namespace JDI_Web.Selenium.Elements.Complex.Table
{
    public enum ElementIndexType { Nums, Names }
    public abstract class TableLine : WebBaseElement, ITableLine
    {
        public bool HasHeader { set; get; }
        public Table Table { get; set; }

        private int _count;

        public int Count
        {
            get => GetCount(false);
            set { if (Table.Cache)
                    _count = value; }
        }
        
        private IList<string> _headers;

        public void AddHeaders(IList<string> headers)
        {
            _headers = headers;
        }
        
        public IList<string> Headers
        {
            set
            {
                if(Table.Cache)
                    _headers = new List<string>(value);
            }
            get
            {
                if (_headers != null)
                    return _headers;
                var localHeaders = HasHeader 
                    ? Timer.GetResult(() => AllHeaders) 
                    : GetNumList(Count);
                if (localHeaders == null || localHeaders.Count == 0)
                    return new List<string>();
                if (Count > 0 && localHeaders.Count > Count)
                {
                    var temp = new List<string>();
                    for (var i = 0; i < Count; i++)
                        temp.Add(localHeaders[i]);
                    localHeaders = temp;
                }
                Headers = localHeaders;
                Count = localHeaders.Count;
                return localHeaders;
            }
        }

        public abstract Dictionary<string, Dictionary<string, ICell>> Get();

        public Dictionary<string, Dictionary<string, string>> AsText {
            get
            {
                return Get().ToDictionary(line => line.Key, line 
                    => line.Value.ToDictionary(el => el.Key, el => el.Value.GetText));
            }
        }
        public Dictionary<string, SelectableElement> Header()
        {
            return GetHeadersAction.ToDictionary(key => key.Text, 
                value => new SelectableElement(webElement: value));
        }

        public SelectableElement Header(string name)
        {
            return Header()[name];
        }

        private IList<string> GetNumList(int count)
        {
            return GetNumList(count, 1);
        }

        private IList<string> GetNumList(int count, int from)
        {
            var result = new List<string>();
            for (var i = from; i < count + from; i++)
               result.Add(i.ToString());
            return result;
        }
        public int StartIndex { set; get; } = 1;
        public By HeadersLocator { set; get; }
        protected By DefaultTemplate { set; get; }
        public By LineTemplate { set; get; }
        public ElementIndexType ElementIndex { set; get; }
        public T Clone<T>(T newLine, Table newTable) where T : TableLine
        {
            newLine.HasHeader = HasHeader;
            newLine.ElementIndex = ElementIndex;
            newLine.Table = newTable;
            newLine.Count = Count;
            newLine.Headers = Headers;
            newLine.StartIndex = StartIndex;
            newLine.HeadersLocator = HeadersLocator;
            newLine.DefaultTemplate = DefaultTemplate;
            newLine.LineTemplate = LineTemplate;
            return newLine;
        }

        public IList<IWebElement> GetLineAction(int colNum)
        {
            return GetElementsByTemplate(colNum + StartIndex - 1);
        }
        protected IList<IWebElement> GetLineAction(string lineName)
        {
            if (LineTemplate != null && LineTemplate.GetByLocator().Contains("{0}"))
                return GetElementsByTemplate(lineName);
            var num = Headers.IndexOf(lineName) + 1; 
            return LineTemplate == null 
                ? GetLineAction(num) 
                : GetElementsByTemplate(num);
        }

        private IList<IWebElement> GetElementsByTemplate(object value)
        {
            var locator = (LineTemplate ?? DefaultTemplate).FillByTemplate(value);
            return Table.WebElement.FindElements(locator).Where(el => el.Displayed).ToList();
        }

        protected abstract IList<IWebElement> GetFirstLine { get; }

        public int GetCount(bool acceptEmpty)
        {
            if (_count > 0)
                return _count;
            if (_headers != null && _headers.Count > 0)
                return _headers.Count;

            var elements = GetHeadersAction;
            if (elements.Count == 0)
                elements = GetFirstLine;
            if (!acceptEmpty)
            {
                var tempElements = elements;
                elements = Timer.GetResultByCondition(() => tempElements, el => el != null && el.Count > 0);
            }
            return elements?.Count(el => el.Displayed) ?? 0;
        }
        public void Clean()
        {
            Headers = null;
            Count = 0;
        }

        public IList<string> AllHeaders => GetHeadersAction.Where(el => el.Displayed)
            .Select(el => el.Text).ToList();
        
        public void Update(TableLine tableLine)
        {
            if (tableLine.Count > 0)
                Count = tableLine.Count;
            if (tableLine.StartIndex != 1)
                StartIndex = tableLine.StartIndex;
            if (tableLine.Headers != null && tableLine.Headers.Count > 0)
                Headers = tableLine.Headers;
            if ((tableLine.GetType().IsAssignableFrom(typeof(Columns)) && !tableLine.HasHeader) ||
                (tableLine.GetType().IsAssignableFrom(typeof(Row)) && tableLine.HasHeader))
                    HasHeader = tableLine.HasHeader;

            if (tableLine.ElementIndex != ElementIndexType.Nums)
                ElementIndex = tableLine.ElementIndex;
        }

        protected IList<IWebElement> GetHeadersAction 
            => Table.WebElement.FindElements(HeadersLocator);
    }
}